### ISSUE TITLE
testng ignores singleThread parameter

### ISSUE THREAD

**Issue by nicholascus** [cbeust/testng#252]
Hi,
I'm running Selenium tests with testNG and have following issue doing parallel test execution.
Execution of test methods from one class could be halted and resumed later instead of using singleThread parameter.

I've attached small demo project which can be built with Maven
https://docs.google.com/open?id=0ByfH7aqUO5g-cml5YTJ2OGRwT28

It includes suite file:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd"> 
<suite verbose="2" name="Connect.Regression"> 
  <test verbose="2" name="all_tests" parallel="classes" thread-count="1"> 
    <classes>
      <class name="au.com.unick.bugz.testngparallel.Test1"/>
      <class name="au.com.unick.bugz.testngparallel.Test2"/>
      <class name="au.com.unick.bugz.testngparallel.Test3"/>
      <class name="au.com.unick.bugz.testngparallel.Test4"/>
      <class name="au.com.unick.bugz.testngparallel.Test5"/>
      <class name="au.com.unick.bugz.testngparallel.Test6"/>
    </classes>
  </test>
</suite>
```

and test classes like this:

```
@Test( singleThreaded=true )
public class Test1 extends GroupOne{
    private static Logger Log = Logger.getLogger(Test1.class.getName());

    @BeforeClass
    public void test1_beforeClass(){
        Log.info(testName+": test1_beforeClass");
    }

    @Test
    public void test1_test1(){
        Log.info(testName+": test1_1");
    }

    @Test(dependsOnMethods={"test1_test1"})
    public void test1_test2(){
        Log.info(testName+": test1_2");
    }
}
```

executing that suite file you can see following output:

```
 T E S T S
-------------------------------------------------------
Running TestSuite
...
... TestNG 6.7 by C?dric Beust (cedric@beust.com)
...

[TestRunner] Starting executor for test all_tests with time out:2147483647 milliseconds.
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:12,995  INFO  Test1: baseTest_beforeClass
[pool-1-thread-1][GroupOne] 27 Jul 2012 13:34:13,002  INFO  Test1: groupOne_beforeClass
[pool-1-thread-1][Test1] 27 Jul 2012 13:34:13,002  INFO  Test1: test1_beforeClass
[pool-1-thread-1][Test1] 27 Jul 2012 13:34:13,005  INFO  Test1: test1_1
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,011  INFO  Test2: baseTest_beforeClass
[pool-1-thread-1][GroupOne] 27 Jul 2012 13:34:13,012  INFO  Test2: groupOne_beforeClass
[pool-1-thread-1][Test2] 27 Jul 2012 13:34:13,012  INFO  Test2: test2_beforeClass
[pool-1-thread-1][Test2] 27 Jul 2012 13:34:13,013  INFO  Test2: test2_1
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,018  INFO  Test3: baseTest_beforeClass
[pool-1-thread-1][GroupTwo] 27 Jul 2012 13:34:13,019  INFO  Test3: groupTwo_beforeClass
[pool-1-thread-1][Test3] 27 Jul 2012 13:34:13,019  INFO  Test3: test3_beforeClass
[pool-1-thread-1][Test3] 27 Jul 2012 13:34:13,020  INFO  Test3: test3_1
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,025  INFO  Test4: baseTest_beforeClass
[pool-1-thread-1][GroupTwo] 27 Jul 2012 13:34:13,026  INFO  Test4: groupTwo_beforeClass
[pool-1-thread-1][Test4] 27 Jul 2012 13:34:13,026  INFO  Test4: test4_beforeClass
[pool-1-thread-1][Test4] 27 Jul 2012 13:34:13,026  INFO  Test4: test4_1
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,032  INFO  Test5: baseTest_beforeClass
[pool-1-thread-1][Test5] 27 Jul 2012 13:34:13,033  INFO  Test5: test5_beforeClass
[pool-1-thread-1][Test5] 27 Jul 2012 13:34:13,033  INFO  Test5: test5_1
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,038  INFO  Test6: baseTest_beforeClass
[pool-1-thread-1][Test6] 27 Jul 2012 13:34:13,039  INFO  Test6: test6_beforeClass
[pool-1-thread-1][Test6] 27 Jul 2012 13:34:13,039  INFO  Test6: test6_1
[pool-1-thread-1][Test1] 27 Jul 2012 13:34:13,041  INFO  Test1: test1_2
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,041  INFO  Test1: baseTest_afterClass
[pool-1-thread-1][Test2] 27 Jul 2012 13:34:13,042  INFO  Test2: test2_2
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,042  INFO  Test2: baseTest_afterClass
[pool-1-thread-1][Test3] 27 Jul 2012 13:34:13,043  INFO  Test3: test3_2
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,043  INFO  Test3: baseTest_afterClass
[pool-1-thread-1][Test4] 27 Jul 2012 13:34:13,043  INFO  Test4: test4_2
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,044  INFO  Test4: baseTest_afterClass
[pool-1-thread-1][Test5] 27 Jul 2012 13:34:13,044  INFO  Test5: test5_2
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,045  INFO  Test5: baseTest_afterClass
[pool-1-thread-1][Test6] 27 Jul 2012 13:34:13,045  INFO  Test6: test6_1
[pool-1-thread-1][BaseTest] 27 Jul 2012 13:34:13,046  INFO  Test6: baseTest_afterClass
```

You can see that Test1 class execution was interrupted after first testing method, then resumed after processing of some other tests.

It's possible to see the same issue with other thread-count limits, you will just have less illustrative test output.

**Comment by krmahadevan** [comment]
[test-ng-test.zip](https://github.com/cbeust/testng/files/1338028/test-ng-test.zip)

Adding the project as a zip, just incase the google drive link stops working.

**Comment by dasAnderl** [comment]
i have the some problem even if tests are executed non parallel. dependson is considered correctly, but methods from different classes get mixed. see the following:

start GEO01_post_def_list
end GEO01_post_def_list
start GEO_NEG01_post_def_list
failed GEO_NEG01_post_def_list
start GEO_NEG01_post_def_list
failed GEO_NEG01_post_def_list
start RSA01_GetSpeedAlertConfig
end RSA01_GetSpeedAlertConfig
start RSA01_GetSpeedAlertConfig
end RSA01_GetSpeedAlertConfig
start RSA02_SaveSpeedAlertDefList
end RSA02_SaveSpeedAlertDefList
start RSA02_SaveSpeedAlertDefList
failed RSA02_SaveSpeedAlertDefList
start RTS01_PostVehicleTripData
end RTS01_PostVehicleTripData
start RTS02_GetTripDataListLongTerm
end RTS02_GetTripDataListLongTerm
start RTS03_GetTripDataListShortTerm
end RTS03_GetTripDataListShortTerm
start RTS04_GetTripDataListCyclic
failed RTS04_GetTripDataListCyclic
start RTS05_GetNewestTripDataLongTerm
end RTS05_GetNewestTripDataLongTerm
start RTS06_GetNewestTripDataShortTerm
end RTS06_GetNewestTripDataShortTerm
start RTS07_GetNewestTripDataCyclic
failed RTS07_GetNewestTripDataCyclic
start RTS08_GetAggregatedTripDataLongTerm
end RTS08_GetAggregatedTripDataLongTerm
start RTS09_GetAggregatedTripDataShortTerm
end RTS09_GetAggregatedTripDataShortTerm
start RTS10_GetAggregatedTripDataCyclic
failed RTS10_GetAggregatedTripDataCyclic
start RTS11_DeleteTripDataLongTerm
end RTS11_DeleteTripDataLongTerm
start RTS12_DeleteTripDataShortTerm
end RTS12_DeleteTripDataShortTerm
start RTS13_DeleteTripDataCyclic
end RTS13_DeleteTripDataCyclic
start RTS14_PostVehicleTripData
end RTS14_PostVehicleTripData
start RTS15_DeleteLongTermTripDataById
end RTS15_DeleteLongTermTripDataById
start RTS16_DeleteShortTermTripDataById
end RTS16_DeleteShortTermTripDataById
start RTS17_DeleteCyclicTripDataById
failed RTS17_DeleteCyclicTripDataById
start RTS18_PostVehicleTripData
end RTS18_PostVehicleTripData
start RTS19_DeleteLongTermTripDataByMileage
end RTS19_DeleteLongTermTripDataByMileage
start RTS20_DeleteShortTermTripDataByMileage
end RTS20_DeleteShortTermTripDataByMileage
start RTS21_DeleteCyclicTripDataByMileage
failed RTS21_DeleteCyclicTripDataByMileage
skipped GEO01_post_def_list
start GEO02_verify_status_is_pending
end GEO02_verify_status_is_pending
skipped GEO_NEG02_verify_status_is_pending
skipped GEO_NEG02_verify_status_is_pending
start RSA03_GetDefListStatus
end RSA03_GetDefListStatus
skipped RSA03_GetDefListStatus
skipped GEO02_verify_status_is_pending
start GEO03_acknowledge_joblist
end GEO03_acknowledge_joblist
skipped GEO_NEG03_acknowledge_joblist
skipped GEO_NEG03_acknowledge_joblist
start RSA04_GetAndAckJob
end RSA04_GetAndAckJob
skipped RSA04_GetAndAckJob
skipped GEO03_acknowledge_joblist
start GEO04_post_JobProcessingCallback_noError
end GEO04_post_JobProcessingCallback_noError
skipped GEO_NEG04_post_JobProcessingCallback_error
skipped GEO_NEG04_post_JobProcessingCallback_error
start RSA05_GetJobProcessingCallback
end RSA05_GetJobProcessingCallback
skipped RSA05_GetJobProcessingCallback
skipped GEO04_post_JobProcessingCallback_noError
start GEO05_verify_status_is_acknowledged
end GEO05_verify_status_is_acknowledged
skipped GEO_NEG05_verify_status_is_error
skipped GEO_NEG05_verify_status_is_error
start RSA06_GetDefListStatus
end RSA06_GetDefListStatus
skipped RSA06_GetDefListStatus
skipped GEO05_verify_status_is_acknowledged
start GEO06_post_violations
end GEO06_post_violations
skipped GEO_NEG06_verify_history_contains_error
skipped GEO_NEG06_verify_history_contains_error
start RSA07_GetSpeedAlertDefinitionList
end RSA07_GetSpeedAlertDefinitionList
skipped RSA07_GetSpeedAlertDefinitionList
skipped GEO06_post_violations
start GEO07_verify_history_contains_activate
end GEO07_verify_history_contains_activate
start RSA08_PostViolation
end RSA08_PostViolation
skipped RSA08_PostViolation
skipped GEO07_verify_history_contains_activate
start GEO08_get_and_verify_alerts
end GEO08_get_and_verify_alerts
start RSA09_GetSpeedAlertHistory
end RSA09_GetSpeedAlertHistory
skipped RSA09_GetSpeedAlertHistory
skipped GEO08_get_and_verify_alerts
start GEO09_get_configuration
end GEO09_get_configuration
start RSA10_GetSpeedAlerts
end RSA10_GetSpeedAlerts
skipped RSA10_GetSpeedAlerts
skipped GEO09_get_configuration
start GEO10_delete_one_alert
end GEO10_delete_one_alert
start RSA11_DeleteSpeedAlertList
end RSA11_DeleteSpeedAlertList
skipped RSA11_DeleteSpeedAlertList
skipped GEO10_delete_one_alert
start GEO11_delete_all_alerts
end GEO11_delete_all_alerts
start RSA12_DeleteRequestHistoryEntriesList
end RSA12_DeleteRequestHistoryEntriesList
skipped RSA12_DeleteRequestHistoryEntriesList
skipped GEO11_delete_all_alerts
start GEO12_delete_one_history_entry
end GEO12_delete_one_history_entry
skipped GEO12_delete_one_history_entry
start GEO13_delete_history
end GEO13_delete_history
skipped GEO13_delete_history

**Comment by adrian-herscu-ge** [comment]
> using class annotation @test(suiteName = "Dwap") on each test fixes the issue

"Dwap" ?? What's this?

**Comment by krmahadevan** [comment]
This is working fine when involving only independent methods (no usage of `dependsOnMethods` and `dependsOnGroups`)

The attached project can be used for cross checking.

When either `dependsOnMethods` (or) `dependsOnGroups` is used, this assurance is broken.

TestNG did introduce a feature called thread affinity wherein downstream methods will run on the same thread as the upstream methods (only when using `dependsOnMethods` and with a single dependency), but even this feature is limited in capabilities.

In short, one shouldn't be relying on thread affinity and expect it to be honored when there are test dependencies involved.

[issue_252.zip](https://github.com/cbeust/testng/files/4332877/issue_252.zip)
